File: backend/app/controllers/email_controller.py
from flask import Blueprint, request, jsonify
from services.email_service import send_authentication_email, send_email_verification, send_email_code, send_rebind_request_email

email_bp = Blueprint("email", __name__)

@email_bp.route("/send-authentication", methods=["POST"])
def send_authentication():
    data = request.json
    username = data.get("username")
    if not username:
        return jsonify({"message": "請輸入用戶名"}), 404
    result = send_authentication_email(username)
    return jsonify(result)

@email_bp.route("/verify-email", methods=["POST"])
def verify_email():
    data = request.json
    username = data.get("username")
    if not username:
        return jsonify({"message": "請輸入用戶名"}), 404
    result = send_email_verification(username)
    if "error" in result:
        return jsonify(result), 400
    return jsonify(result), 200

@email_bp.route("/verify-code", methods=["POST"])
def verify_code():
    data = request.json
    if not data:
        return jsonify({"error": "未收到請求資料"}), 400

    username = data.get("username")
    verification_code = data.get("verificationCode")
    if not username:
        return jsonify({"error": "請輸入用戶名"}), 404
    if not verification_code:
        return jsonify({"error": "請輸入信箱驗證碼"}), 404
    result = send_email_code(username, verification_code)
    if "error" in result:
        return jsonify(result), 400
    return jsonify(result)

@email_bp.route("/request-bind-email", methods=["POST"])
def request_rebind_email():
    data = request.get_json()
    username = data.get("username")

    if not username:
        return jsonify({"message": "缺少 username"}), 400
    try:
        result = send_rebind_request_email(username)
        return jsonify(result)
    except Exception as e:
        return jsonify({"message": "寄信失敗", "error": str(e)}), 500

File: backend/golang-app/controllers/email_controller.go
package controllers

import (
	"net/http"

	"golang-app/models"
	"golang-app/services"

	"github.com/gin-gonic/gin"
)

func RegisterEmailRoutes(r *gin.Engine) {
	r.POST("/send-authentication", SendAuthentication)
	r.POST("/verify-email", VerifyEmail)
	r.POST("/verify-code", VerifyCode)
	r.POST("/request-bind-email", RequestRebindEmail)
}

func SendAuthentication(c *gin.Context) {
	var req struct {
		Username string `json:"username"`
	}

	if err := c.ShouldBindJSON(&req); err != nil || req.Username == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "請輸入用戶名"})
		return
	}

	result := services.SendAuthenticationEmail(req.Username, models.DB)
	c.JSON(http.StatusOK, result)
}

func VerifyEmail(c *gin.Context) {
	var req struct {
		Username string `json:"username"`
	}

	if err := c.ShouldBindJSON(&req); err != nil || req.Username == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "請輸入用戶名"})
		return
	}

	result := services.SendEmailVerification(req.Username, models.DB)
	if errMsg, ok := result["error"]; ok {
		c.JSON(http.StatusBadRequest, gin.H{"error": errMsg})
		return
	}

	c.JSON(http.StatusOK, result)
}

func VerifyCode(c *gin.Context) {
	var req struct {
		Username         string `json:"username"`
		VerificationCode string `json:"verificationCode"`
	}

	if err := c.ShouldBindJSON(&req); err != nil || req.Username == "" || req.VerificationCode == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "缺少用戶名或驗證碼"})
		return
	}

	result := services.SendEmailCode(req.Username, models.DB, req.VerificationCode)
	if errMsg, ok := result["error"]; ok {
		c.JSON(http.StatusBadRequest, gin.H{"error": errMsg})
		return
	}

	c.JSON(http.StatusOK, result)
}

func RequestRebindEmail(c *gin.Context) {
	var req struct {
		Username string `json:"username"`
	}

	if err := c.ShouldBindJSON(&req); err != nil || req.Username == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "缺少 username"})
		return
	}

	result, err := services.SendRebindRequestEmail(req.Username, models.DB)
	if err != 200 {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "寄信失敗", "detail": err})
		return
	}

	c.JSON(http.StatusOK, result)
}
